crates/protocol/src/messages/events.rs

```diff
@@ -0,0 +1,135 @@
+//! Message event primitives for OP stack interoperability.
+//!
+//! <https://specs.optimism.io/interop/messaging.html#messaging>
+//! <https://github.com/ethereum-optimism/optimism/blob/34d5f66ade24bd1f3ce4ce7c0a6cfc1a6540eca1/packages/contracts-bedrock/src/L2/CrossL2Inbox.sol>
+use alloc::vec;
+use alloy_primitives::{keccak256, Address, Bytes, Log, U256};
+use alloy_sol_types::{sol, SolType};
+use derive_more::{AsRef, From};
+
+sol! {
+    /// @notice The struct for a pointer to a message payload in a remote (or local) chain.
+    #[derive(Default, Debug, PartialEq, Eq)]
+    struct MessageIdentifierAbi {
+        address origin;
+        uint256 blockNumber;
+        uint256 logIndex;
+        uint256 timestamp;
+        uint256 chainId;
+    }
+
+    /// @notice Emitted when a cross chain message is being executed.
+    /// @param msgHash Hash of message payload being executed.
+    /// @param id Encoded Identifier of the message.
+    #[derive(Default, Debug, PartialEq, Eq)]
+    event ExecutingMessage(bytes32 indexed msgHash, MessageIdentifierAbi id);
+
+    /// @notice Executes a cross chain message on the destination chain.
+    /// @param _id      Identifier of the message.
+    /// @param _target  Target address to call.
+    /// @param _message Message payload to call target with.
+    function executeMessage(
+        MessageIdentifierAbi calldata _id,
+        address _target,
+        bytes calldata _message
+    ) external;
+}
+
+/// A [`MessagePayload`] is the raw payload of an initiating message.
+#[derive(Debug, Clone, From, AsRef, PartialEq, Eq)]
+pub struct MessagePayload(Bytes);
+
+impl From<Log> for MessagePayload {
+    fn from(log: Log) -> Self {
+        let mut data = vec![0u8; log.topics().len() * 32 + log.data.data.len()];
+        for (i, topic) in log.topics().iter().enumerate() {
+            data[i * 32..(i + 1) * 32].copy_from_slice(topic.as_ref());
+        }
+        data[(log.topics().len() * 32)..].copy_from_slice(log.data.data.as_ref());
+        Bytes::from(data).into()
+    }
+}
+
+/// A [`MessageIdentifier`] uniquely represents a log that is emitted from a chain within
+/// the broader dependency set. It is included in the calldata of a transaction sent to the
+/// CrossL2Inbox contract.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "serde", serde(rename_all = "camelCase"))]
+pub struct MessageIdentifier {
+    /// The account that sent the message.
+    pub origin: Address,
+    /// The block number that the message was sent in.
+    pub block_number: u64,
+    /// The log index of the message in the block (global).
+    pub log_index: u64,
+    /// The timestamp of the message.
+    pub timestamp: u64,
+    /// The chain ID of the chain that the message was sent on.
+    #[cfg_attr(feature = "serde", serde(rename = "chainID"))]
```

Comment:
not sure if this is `chainId` or `chainID`

Reply:
`chainID`, see [op-geth](https://github.com/ethereum-optimism/op-geth/blob/optimism/core/types/interoptypes/interop.go#L101)